src/lib/gadgets/basic.ts

```diff
@@ -0,0 +1,79 @@
+/**
+ * Basic gadgets that only use generic gates
+ */
+import type { Field, VarField } from '../field.js';
+import { existsOne, toVar } from './common.js';
+import { Gates } from '../gates.js';
+import { TupleN } from '../util/types.js';
+
+export { assertOneOf };
+
+// TODO: create constant versions of these and expose on Gadgets
+
+/**
+ * Assert that a value equals one of a finite list of constants:
+ * `(x - c1)*(x - c2)*...*(x - cn) === 0`
+ *
+ * TODO: what prevents us from getting the same efficiency with snarky DSL code?
+ */
+function assertOneOf(x: Field, allowed: [bigint, bigint, ...bigint[]]) {
```

Comment:
What does this notation `... bigint[]` mean?

Reply:
Is it like a tail of a list with potentially more bigints?

Reply:
yes! I made the type require at least two inputs so that I don't have to implement the case of 0 or 1 inputs